heudiconv/tests/test_utils.py

```diff
@@ -173,19 +176,85 @@ def test_get_datetime() -> None:
+        ("2023-04-02T11:47:09.000000", "%Y-%m-%dT%H:%M:%S.%f"),
+        ("2023-04-02T11:47:09.1", "%Y-%m-%dT%H:%M:%S.%f"),
+        ("2023-04-02T11:47:09-0900", "%Y-%m-%dT%H:%M:%S%z"),
+        ("2023-04-02T11:47:09.1-0900", "%Y-%m-%dT%H:%M:%S.%f%z"),
     ],
 )
-def test_strptime_micr(dt: str, fmt: str) -> None:
+def test_strptime_bids(dt: str, fmt: str) -> None:
     target = datetime.strptime(dt, fmt)
-    assert strptime_micr(dt, fmt) == target
-    assert strptime_micr(dt, fmt + "[.%f]") == target
-    assert strptime_micr(dt + ".0", fmt + "[.%f]") == target
-    assert strptime_micr(dt + ".000000", fmt + "[.%f]") == target
-    assert strptime_micr(dt + ".1", fmt + "[.%f]") == datetime.strptime(
-        dt + ".1", fmt + ".%f"
-    )
+    assert strptime_bids(dt) == target
+
+
+@pytest.mark.parametrize(
+    "tm, tm_fmt",
+    [
+        ("114709.1", "%H%M%S.%f"),
+        ("114709", "%H%M%S"),
+        ("1147", "%H%M"),
+        ("11", "%H"),
+    ],
+)
+@pytest.mark.parametrize(
+    "offset, offset_fmt",
+    [
+        ("-0900", "%z"),
+        ('', ''),
+    ],
+)
+def test_strptime_dcm_da_tm(tm: str, tm_fmt: str, offset: str, offset_fmt: str) -> None:
+    da = "20230402"
+    da_fmt = "%Y%m%d"
+    target = datetime.strptime(da + tm + offset, da_fmt + tm_fmt + offset_fmt)
+    ds = dcm.dataset.Dataset()
+    ds["AcquisitionDate"] = dcm.DataElement("AcquisitionDate","DA",da)
+    ds["AcquisitionTime"] = dcm.DataElement("AcquisitionTime", "TM", tm)
+    if offset:
+        ds[(0x0008, 0x0201)] = dcm.DataElement((0x0008, 0x0201), "SH", offset)
+    assert strptime_dcm_da_tm(ds, "AcquisitionDate", "AcquisitionTime") == target
+
+
+@pytest.mark.parametrize(
+    "dt, dt_fmt",
+    [
+        ("20230402114709.1-0400", "%Y%m%d%H%M%S.%f%z"),
+        ("20230402114709-0400", "%Y%m%d%H%M%S%z"),
+        ("202304021147-0400", "%Y%m%d%H%M%z"),
+        ("2023040211-0400", "%Y%m%d%H%z"),
+        ("20230402-0400", "%Y%m%d%z"),
+        ("202304-0400", "%Y%m%z"),
+        ("2023-0400", "%Y%z"),
+        ("20230402114709.1", "%Y%m%d%H%M%S.%f"),
+        ("20230402114709", "%Y%m%d%H%M%S"),
+        ("202304021147", "%Y%m%d%H%M"),
+        ("2023040211", "%Y%m%d%H"),
+        ("20230402", "%Y%m%d"),
+        ("202304", "%Y%m"),
+        ("2023", "%Y"),
+    ],
+)
+@pytest.mark.parametrize(
+    "offset, offset_fmt",
+    [
+        ("-0900", "%z"),
+        ('', ''),
+    ],
+)
+def test_strptime_dcm_dt(dt: str, dt_fmt: str, offset: str, offset_fmt: str) -> None:
+    target = None
+    if dt_fmt[-2:] == "%z" and offset:
+        target = datetime.strptime(dt, dt_fmt)
+    else:
+        target = datetime.strptime(dt + offset, dt_fmt + offset_fmt)
+    ds = dcm.dataset.Dataset()
+    ds["AcquisitionDateTime"] = dcm.DataElement("AcquisitionDateTime","DT", dt)
+    if offset:
+        ds[(0x0008, 0x0201)] = dcm.DataElement((0x0008, 0x0201), "SH", offset)
+    assert strptime_dcm_dt(ds, "AcquisitionDateTime") == target
```

Comment:
It is unfortunate that we do not carry any "proper" dicom file with timezone info. But FWIW our dependencies do:

```
❯ /usr/bin/find -iname *.dcm | while read d; do off=$(dcmdump +P 0008,0201 $d 2>/dev/null) ; [ -n "$off" ] || continue; echo -e "$d\t$off"; done | head
./.tox/typing/lib/python3.11/site-packages/pydicom/data/charset_files/chrArab.dcm	(0008,0201) SH [-0400]                                  #   6, 1 TimezoneOffsetFromUTC
./.tox/typing/lib/python3.11/site-packages/pydicom/data/charset_files/chrFren.dcm	(0008,0201) SH [-0400]                                  #   6, 1 TimezoneOffsetFromUTC
./.tox/typing/lib/python3.11/site-packages/pydicom/data/charset_files/chrFrenMulti.dcm	(0008,0201) SH [-0400]                                  #   6, 1 TimezoneOffsetFromUTC

```

but I do not think we should rely on them.  But we could copy some I guess. WDYT, worth it?

Reply:
I'm not sure I see the benefit to adding dicom files to the git repository. Generating the dicom from scratch facilitates testing every tag allowed by the standard without keeping numerous files in the repository.

Reply:
there are pros/cons for both approaches. e.g. for existing/committed (to git or via git-annex) files -- ensuring that we are operating correctly on the files as provided by manufacturers. Not surprising that dcm2niix has all those collections linked as submodules: https://github.com/rordenlab/dcm2niix/blob/master/.gitmodules

Reply:
but you are correct that on the fly generation could be quite handy for how you have done them above - in case of parametric tests etc. I was just wondering if we could add testing on some existing file with such metadata in it... but if none there -- let's forget about this particular aspect